Change import

ElMartes opened the pull request:
Cleaner initialization of the metro module

philsmt committed:
Move devices module to services package to avoid circular import with mp spawn method

ElMartes committed:
Added dependency on xarray module

philsmt committed:
Add metro.init_mp_support to fix missing initialization with spawn method

philsmt committed:
No longer import metro.services.devices into the metro namespace as a module, but add all public symbols

philsmt committed:
Rename GenericDevice.__ge__ to isSubDevice and simplify the implementation

ElMartes committed:
Restructures the initialization routines in the metro module for clearer importing

ElMartes committed:
Restructures the initialization routines in the metro module for clearer importing

ElMartes committed:
Change global flag load_GUI

ElMartes committed:
Change global flag load_GUI

ElMartes committed:
Fix Python version checking and 'die' routine initialization

ElMartes committed:
Fix Python version checking and 'die' routine initialization

ElMartes committed:
Remove obsolete comments

ElMartes committed:
Remove obsolete comments

philsmt commented:
Thanks for this attempt to tackle the package import conundrum.

Before we go into the technical details, please rebase this work on top of the current master with #2 merged now to make the diff more manageable. Furthermore, please run flake8 on your code. It _will_ complain about a few formatting choices. For convenience, I updated the `util/flake8.sh` script for the latest version and applied its newest complaints.

Finally, the non capitalized command line descriptions are on purpose to be in line with the builtin entries created by `argparse` itself.

ElMartes committed:
Style adjustments for flake8 and consistence

ElMartes committed:
Style adjustments for flake8 and consistence

ElMartes committed:
Merge branch 'change_import' of https://github.com/philsmt/metro-sci into change_import

ElMartes commented:
Rebased onto the master branch.
I changed the style according to the flake8 complaints, but there are still some complaints with other files not altered in this branch.
As for the capitalized descriptions: I changed them because they are formulated as full sentences including a period at the end, but I now instead removed the period, in accordance with the usual argparse style.

ElMartes committed:
Add source root path of the metro package to the globals

ElMartes committed:
Fix taskbar icon in Windows

ElMartes committed:
Fix problem with loading invisible devices from a profile

ElMartes committed:
Merge remote-tracking branch 'origin/master' into change_import

philsmt commented:
Got EXtra-metro working with a minor name and argument change related to the `init` function. Indeed this interface makes sense, I'll double-check later why I ever passed the `args` object directly. Any particular reason you wanted to rename `init` to `init_metro` here?

Could you please revert the various style changes not relevant for the functionality change? This will also make the diff clearer.

ElMartes committed:
Revert style changes in __init__ code, fix minor import bug, rename initialization routine back to 'init'

ElMartes committed:
Fix parallel_operator to account for renaming of initialization routine

ElMartes commented:
I reverted the style changes and fixed some minor importing issue.
Also I renamed the main initialization routine back to "init", which can be called via, e.g., metro.init(args.core_mode, args.kiosk_mode, window_title).

philsmt commented on src/metro/services/devices.py:
```diff
@@ -383,6 +383,8 @@ def _prepare(self, name, parent, args, state, entry_point):
             self.destroyed.connect(_on_device_destroyed)
 
+        # for correct/full initialization set as visible first
+        self.setVisible(True)
```
What specific problem does this solve? It seems not ideal to potentially pop up a window briefly, just to hide it immediately.

ElMartes replied:
It solves a bug where a profile could not be loaded when the device is directly initialized as invisible (i.e., when the corresponding window of the device was hidden upon saving the profile). Took me a while to find, but it seems that invisible devices are somehow not properly initialized by Qt which leads to problems.
Maybe this bug is solved in a newer version of Qt, however, as this is already a couple years old by now.

philsmt replied:
Thanks for the clarification, let's keep it for now. I didn't notice any weird flickering behaviour when using it.

philsmt reviewed (approved):
As you may have noticed, we've been working on getting `metro-sci` compatible with modern versions of Python and Qt in `fix/forward-compat`. This is fairly critical for us to continue supporting it on modern hardware.

I'd love to get this done finally and most of the testing on it already. As its history is unfortunately a bit wonky by mergers of masters into it, I did not manage to rebase it. while merging into it seems to work. I would ask you to squeeze the commits please to get rid of the duplicate commits.
